### ISSUE TITLE
Don't set global workqueue MetricsProvider

### ISSUE THREAD

**Issue by austince** [kubernetes-sigs/controller-runtime#2238]
Currently, controller-runtime [sets the global workqueue MetricsProvider](https://github.com/kubernetes-sigs/controller-runtime/blob/e2d8821b277fb57f3a951f430bb1746d8297b853/pkg/metrics/workqueue.go#L99) in an init method which makes it impossible for users of the library to configure their own.

As noted in:
* https://github.com/kubernetes/client-go/issues/1193
* https://github.com/kubernetes/kubernetes/pull/114242#issuecomment-1345613392
* SIG API meeting: https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit?pli=1#heading=h.41qdxxsr3sdz

**Comment by alvaroaleman** [comment]
I thought about this issue the other day and one possible approach that came to mind is that rather than using an `init()`, we could register these the first time a controller is created. This would:
* Not break any existing user who wants to just keep using the controller-runtime metrics
* Allow advanced users to set their own provider by doing that before they create a controller

**Comment by alvaroaleman** [comment]
@sbueringer made me aware that this might break a lot of existing users, because component-base also does this in an init and today controller-runtime happens to "win": https://github.com/kubernetes/component-base/blob/3b0f2cefb281ec6078f2949fe7041619ddcfd262/metrics/prometheus/workqueue/metrics.go#L104

**Comment by JoelSpeed** [comment]
Per a bug I was investigating this morning, you may not always win that race. Depending on your import graph, you may end up pulling in component-base earlier than controller-runtime and therefore, it executes the component-base version of this prior to the controller-runtime version, which means we lost.

A non-breaking way to do this would be to expose the metrics provider as a public object. I believe users now have access to create their own `NewQueue` implementation, which means they could pass the metrics provider into that. But currently, the metrics provider is all private. If this were public, could I not then explicitly pass it, causing the queue to leverage this provider over the global provider and avoiding any races?

**Comment by alvaroaleman** [comment]
Hmhm. I am wondering if rather than setting the global metric provider for the queues, we could set it only in the opts for the queues we construct if not already set? Then things will keep working for everyone without us having to snatch the global provider. The only case I can think of where that would cause issues is if someone constructs a queue themselves and relies on its metrics shown up in controller-runtime.

**Comment by JoelSpeed** [comment]
>The only case I can think of where that would cause issues is if someone constructs a queue themselves and relies on its metrics shown up in controller-runtime.

**Comment by alvaroaleman** [comment]
> In that case is it not just a case of allowing them to pass the metrics provider in the queue options?

Yeah but it still would break them by default. But yeah, making the provider public so that can be done is likely good enough.

**Comment by xuzhenglun** [comment]
As a quick fix, maybe we can expose `MetricProvider` as configurable? As the following code, no matter priority queue or the classic one, `MetricProvider` is not passed from `Opts` and using the default one. 
https://github.com/kubernetes-sigs/controller-runtime/blob/d2c0c8b55aed8a473b23d4a860103e7d096377be/pkg/controller/controller.go#L262-L270

Since implicit behavior is difficult to change, perhaps allowing users to explicitly specify a choice would be a more realistic option?
